rct229/rulesets/ashrae9012019/section19/section19rule21.py

```diff
@@ -0,0 +1,498 @@
+                        ] = False
+
+                    for hvac_id_p in get_list_hvac_systems_associated_with_zone(
+                        rmd_p, zone_id_b
+                    ):
+                        hvac_p = find_one(
+                            f'$.buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*][?(@.id = "{hvac_id_p}")]',
+                            rmd_p,
+                        )
+                        ER_modeled_p = find_one(
+                            "$.fan_system.air_energy_recovery.energy_recovery_type",
+                            hvac_p,
+                        ) in (
+                            ENERGY_RECOVERY.SENSIBLE_HEAT_EXHANGE,
+                            ENERGY_RECOVERY.ENTHALPY_HEAT_EXHANGE,
+                            ENERGY_RECOVERY.SENSIBLE_HEAT_WHEEL,
+                            ENERGY_RECOVERY.ENTHALPY_HEAT_WHEEL,
+                            ENERGY_RECOVERY.HEAT_PIPE,
+                            ENERGY_RECOVERY.OTHER,
+                        )
+
+                        serves_zones_that_have_dehumid_heat_recovery = (
+                            find_one("$.cooling_system.dehumidification_type", hvac_p)
+                            == DEHUMIDIFICATION.SERIES_HEAT_RECOVERY
+                        )
+
+                    hvac_systems_and_assoc_zones_largest_exhaust_source[
+                        hvac_id_b
+                    ] = get_hvac_sys_and_assoc_zones_largest_exhaust_source(
+                        rmd_b, hvac_id_b
+                    )
+
+                sys_type_heating_only = any(
+                    [
+                        hvac_id_b in baseline_hvac_system_dict[applicable_sys_type]
+                        for applicable_sys_type in APPLICABLE_SYS_TYPES
+                    ]
+                )
+
+            try:
+                max_thermostat_heating_setpoint_schedule_p = max(
+                    [
+                        max(thermostat_schedule)
+                        * ureg(
+                            "degC"
+                        )  # TODO: make sure this unit is degC. The current schema (v 0.0.29) doesn't specify the unit specifically
+                        for thermostat_schedule_id in find_all(
+                            "$.buildings[*].building_segments[*].zones[*].thermostat_heating_setpoint_schedule",
+                            rmd_p,
+                        )
+                        for thermostat_schedule in find_all(
+                            f'$.schedules[*][?(@.id = "{thermostat_schedule_id}")].hourly_values',
+                            rmd_p,
+                        )
+                    ]
+                )
+            except ValueError:
+                max_thermostat_heating_setpoint_schedule_p = ZERO.TEMPERATURE
+
+            serves_zones_heated_to_60_or_higher_p = (
+                max_thermostat_heating_setpoint_schedule_p > REQ_HEATING_SETPOINT
+            )
```

Comment:
I think this should be added to the zone_data, instead of a stand-alone data.

Reply:
Addressed.